File: py/assemble.py
import numpy as np
from dataclasses import dataclass


@dataclass
class DofMap:
    n_nodes: int
    ndofs_per_node: int = 1

    @property
    def n_dofs(self) -> int:
        return self.n_nodes * self.ndofs_per_node

    def element_dofs(self, conn_e: np.ndarray) -> np.ndarray:
        if self.ndofs_per_node == 1:
            return conn_e


def assemble_poisson(mesh, ref_el, kernel, quad_order: int = 1, cell_type: str = "triangle"):
    conn = mesh.cells[cell_type]
    n_elems = conn.shape[0]
    nen = conn.shape[1]

    n_nodes = mesh.coords.shape[0]

    K = np.zeros((n_nodes, n_nodes))
    F = np.zeros(n_nodes)

    for e in range(n_elems):
        nodes = conn[e]
        coords_e = mesh.coords[nodes]

        Ke, Fe = kernel.element_matrix(coords_e, ref_el, quad_order)

        for a in range(nen):
            A = nodes[a]
            F[A] += Fe[a]

            for b in range(nen):
                B = nodes[b]
                K[A, B] += Ke[a, b]

    return K, F

File: assemble.cpp
#include "assemble.hpp"

Assemble::GlobalData Assemble::assemble_poisson(const Mesh2d& mesh, PoissonKernel& kernel)
{
    const auto&       conn    = mesh.triangles;
    const std::size_t n_elems = conn.size();
    const std::size_t nen     = conn[0].size();
    const std::size_t n_nodes = mesh.coords.size();
    GlobalData        out     = {};

    if (n_elems == 0)
    {
        return out;
    }

    out.K.resize(n_nodes);

    for (std::size_t i = 0; i < n_nodes; ++i)
    {
        out.K[i].resize(n_nodes);
    }
    out.F.resize(n_nodes);

    for (std::size_t i = 0; i < n_elems; ++i)
    {
        const auto&                  nodes = conn[i];
        PoissonKernel::ElementPoints coords_e;

        for (std::size_t j = 0; j < nen; ++j)
        {
            coords_e[j] = mesh.coords[nodes[j]];
        }

        PoissonKernel::ElementData data = kernel.element_matrix(coords_e);

        for (std::size_t a = 0; a < nen; ++a)
        {
            std::size_t A = nodes[a];
            out.F[A] += data.Fe[a];

            for (std::size_t b = 0; b < nen; ++b)
            {
                std::size_t B = nodes[b];
                out.K[A][B] += data.Ke[a][b];
            }
        }
    }

    return out;
}